okhttp/src/main/java/okhttp3/CertificatePinner.kt

```diff
@@ -0,0 +1,339 @@
+ *     .add("publicobject.com", "sha256/grX4Ta9HpZx6tSHkmCrvpApTQGo67CYDnvprLg5yRME=")
+ *     .add("publicobject.com", "sha256/lCppFqbkrlJ3EcVFAkeip0+44VaoJUymbnOaEUk7tEU=")
+ *     .build();
+ * ```
+ *
+ * Pinning is per-hostname and/or per-wildcard pattern. To pin both `publicobject.com` and
+ * `www.publicobject.com`, you must configure both hostnames.
+ *
+ * Wildcard pattern rules:
+ *
+ *  1. Asterisk `*` is only permitted in the left-most domain name label and must be the only
+ *     character in that label (i.e., must match the whole left-most label). For example,
+ *     `*.example.com` is permitted, while `*a.example.com`, `a*.example.com`, `a*b.example.com`,
+ *     `a.*.example.com` are not permitted.
+ *  2. Asterisk `*` cannot match across domain name labels. For example, `*.example.com` matches
+ *     `test.example.com` but does not match `sub.test.example.com`.
+ *  3. Wildcard patterns for single-label domain names are not permitted.
+ *
+ * If hostname pinned directly and via wildcard pattern, both direct and wildcard pins will be used.
+ * For example: `*.example.com` pinned with `pin1` and `a.example.com` pinned with `pin2`, to check
+ * `a.example.com` both `pin1` and `pin2` will be used.
+ *
+ * Warning: Certificate Pinning is Dangerous!
+ * ------------------------------------------
+ *
+ * Pinning certificates limits your server team's abilities to update their TLS certificates. By
+ * pinning certificates, you take on additional operational complexity and limit your ability to
+ * migrate between certificate authorities. Do not use certificate pinning without the blessing of
+ * your server's TLS administrator!
+ *
+ * ### Note about self-signed certificates
+ *
+ * [CertificatePinner] can not be used to pin self-signed certificate if such certificate is not
+ * accepted by [javax.net.ssl.TrustManager].
+ *
+ * @see [OWASP: Certificate and Public Key
+ *     Pinning](https://www.owasp.org/index.php/Certificate_and_Public_Key_Pinning)
+ */
+class CertificatePinner internal constructor(
+  private val pins: Set<Pin>,
+  private val certificateChainCleaner: CertificateChainCleaner?
+) {
+  /**
+   * Confirms that at least one of the certificates pinned for `hostname` is in `peerCertificates`.
+   * Does nothing if there are no certificates pinned for `hostname`. OkHttp calls this after a
+   * successful TLS handshake, but before the connection is used.
+   *
+   * @throws SSLPeerUnverifiedException if `peerCertificates` don't match the certificates pinned
+   *     for `hostname`.
+   */
+  @Throws(SSLPeerUnverifiedException::class)
+  fun check(hostname: String, peerCertificates: List<Certificate>) {
+    var peerCertificates = peerCertificates
+    val pins = findMatchingPins(hostname)
+    if (pins.isEmpty()) return
+
+    if (certificateChainCleaner != null) {
+      peerCertificates = certificateChainCleaner.clean(peerCertificates, hostname)
+    }
+
+    for (c in 0 until peerCertificates.size) {
```

Comment:
peerCertificates.foreach?

Reply:
or peerCertificates.find{} ?: throw AssertionError("unsupported hashAlgorithm: ${pin.hashAlgorithm}")
or peerCertificates.any{} || throw AssertionError("unsupported hashAlgorithm: ${pin.hashAlgorithm}")

Reply:
Switched to foreach. No strong feelings about find either way.